### DISCUSSION TITLE
Using rich.traceback with traceback.print_stack()

### DISCUSSION THREAD

**Discussion by NTL-Remi** [Textualize/rich#1531] [Q&A]
Hello,
I would like to know if there is a way to format the output of the standard library's [`traceback.print_stack()`](https://docs.python.org/3/library/traceback.html#traceback.print_stack) using[ `rich.traceback`](https://rich.readthedocs.io/en/latest/traceback.html).

`traceback.print_stack()` displays a traceback where it is called, without interrupting the program,
but from the docs, it seems that `rich.traceback` needs to raise an exception to work.

Is there any workaround to pretty print a trace without ending a program ?
I would like to use it to pinpoint warnings, following the first part of [this Stack Overflow answer](https://stackoverflow.com/a/22376126/17035199).

**Comment by willmcgugan** [comment]
Please see Printing Tracebacks section.

https://rich.readthedocs.io/en/stable/traceback.html

**Reply by NTL-Remi** [reply]
I already did, but it does not really help.
I tried the following, combining the `Printing Tracebacks` section examples and the `warnings.showwarning` reassignment proposed in the Stack Overflow answer:
```python
import warnings

import rich.console


def showwarning(message, *args, **kwargs):
    try:
        raise Exception
    except Exception:
        rich.console.Console().print_exception()


warnings.showwarning = showwarning


def test_function():
    warnings.warn('test')


test_function()
```

```
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ <stdin>:3 in showwarning                                                                         │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
Exception
```

It doesn't work well, there is only one frame in the trace, the one of `showwarning`, nothing about test_function().
I tried in more complex code, and it was the same, only one frame.

Using instead
```python
import traceback


def showwarning(message, *args, **kwargs):
    traceback.print_stack()
```
correctly outputs all the frames
```
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in test_function
  File "/home/.../python3.9/warnings.py", line 109, in _showwarnmsg
    sw(msg.message, msg.category, msg.filename, msg.lineno,
  File "<stdin>", line 2, in showwarning
```

**Reply by willmcgugan** [reply]
`print_exception` will only print a trace for exceptions. There's currently no way to just print the stack.

**Reply by NTL-Remi** [reply]
Ok, thank you, that was my question.
So there is no workaround.

**Reply by willmcgugan** [reply]
It's not supported currently. You are the first to ask for it.

**Reply by NTL-Remi** [reply]
Yes I guess it is not something a lot of people would use.
I mostly use `rich` to format logs from `loguru`, that I also use to redirect warnings to logs by using the showwarning trick.
But by doing this, warnings now emanate from the showwarning function, so all the information I have are the warning messages.
Having the stack trace is useful, especially when the warnings come from dependency libraries.

**Reply by toppk** [reply]
I'd like that too.  This is a function I created so I can sweet rich.traceback output whenever I want.

```python
def pretty_trace():
    from rich.console import Console
    from rich.traceback import Traceback
    from sys import _getframe
    from types import TracebackType
    tb = None
    depth = 1
    while True:
        try:
            frame = _getframe(depth)
            depth += 1
        except ValueError as exc:
            break
        tb = TracebackType(tb, frame, frame.f_lasti, frame.f_lineno)
    ex = Exception('pretty_trace')
    stack = Traceback.extract(type(ex), ex, tb, show_locals=True)
    Console().print(Traceback(stack, show_locals=True))
```
